blockchain/spam_throttler.go

```diff
@@ -0,0 +1,304 @@
+			newThreshold = t.threshold - t.config.ThresholdAdjustment
+		} else {
+			// Set minimum threshold
+			newThreshold = t.config.MinimumThreshold
+		}
+
+		// Increase threshold if a fail ratio is smaller than target ratio until it exceeds InitialThreshold
+	} else {
+		if t.threshold+t.config.ThresholdAdjustment < t.config.InitialThreshold {
+			newThreshold = t.threshold + t.config.ThresholdAdjustment
+		} else {
+			// Set maximum threshold
+			newThreshold = t.config.InitialThreshold
+		}
+	}
+
+	t.threshold = newThreshold
+
+	// Update metrics
+	thresholdGauge.Update(int64(newThreshold))
+}
+
+// newAllowed generates a new allowed list of throttler.
+func (t *throttler) newAllowed(allowed []common.Address) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	a := make(map[common.Address]bool, len(allowed))
+	for _, addr := range allowed {
+		a[addr] = true
+	}
+	t.allowed = a
+}
+
+// updateThrottled removes outdated addresses from the throttle list and adds new addresses to the list.
+func (t *throttler) updateThrottled(newThrottled []common.Address) {
+	var removeThrottled []common.Address
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// Decrease throttling remained time for all throttled addresses.
+	for addr, remained := range t.throttled {
+		t.throttled[addr] = remained - 1
+		if t.throttled[addr] < 0 {
+			removeThrottled = append(removeThrottled, addr)
+		}
+	}
+
+	// Remove throttled addresses from throttled map.
+	for _, addr := range removeThrottled {
+		delete(t.throttled, addr)
+	}
+
+	for _, addr := range newThrottled {
+		t.throttled[addr] = t.config.ThrottleSeconds
+	}
+
+	// Update metrics
+	throttledSizeGauge.Update(int64(len(t.throttled)))
+	allowedSizeGauge.Update(int64(len(t.allowed)))
+}
+
+// updateThrottlerState updates the throttle list by calculating spam weight of candidates.
+func (t *throttler) updateThrottlerState(txs types.Transactions, receipts types.Receipts) {
+	var removeCandidate []common.Address
+	var newThrottled []common.Address
+
+	startTime := time.Now()
+	numFailed := 0
+	failRatio := uint(0)
+	mapSize := uint(len(t.candidates))
+
+	// Increase spam weight of throttle candidates who generate failed txs.
+	for i, receipt := range receipts {
+		if receipt.Status != types.ReceiptStatusSuccessful {
+			numFailed++
+
+			toAddr := txs[i].To()
+			if toAddr == nil {
+				continue
+			}
+
+			weight := t.candidates[*toAddr]
+			if weight == 0 {
+				if mapSize >= t.config.MaxCandidates {
+					continue
+				}
+				mapSize++
+			}
+
+			t.candidates[*toAddr] = weight + t.config.IncreaseWeight
+		}
+	}
+
+	// Decrease spam weight for all candidates and update throttle lists in throttled.
+	for addr, weight := range t.candidates {
+		newWeight := weight - t.config.DecreaseWeight
+
+		switch {
+		case newWeight <= 0:
+			removeCandidate = append(removeCandidate, addr)
+
+		case newWeight > t.threshold:
+			removeCandidate = append(removeCandidate, addr)
```

Comment:
@aidan-kwon 
Hi aidan. Sorry for late review, but have a question here.

Does `newWeight > t.threshold` means "it should be limited because it exceeds threshold"? 
If so, I can understand Line 218, but hard to grasp meaning of Line 217 because I think `removeCandidate` means "the address list which should be removed because now it is under threshold."

Can you explain why Line 217 is needed? 
Thanks in advance.

Reply:
1) For the ease of managing `candidates` and `throttled`
Both threshold and weight are variable. Even though a weight keeps decreasing, it can be re-registered in the throttled list if the threshold decrease more sharply. It could be false-positive detecting of throttled address. And, decreasing weight to 0 takes many times (100 ~ 500 blocks). The throttle takes lots of time to manage weights of already throttled addresses

2) For continuous throttling 
Even though an address is in the `throttled`, it may keep generating revert txs exceeding the threshold. In that case, current logic allows resetting `ThrottleSeconds` of the throttled address by accumulating the weight again